Regenerate topic base on connect

amoss opened the pull request:
##### Summary
Closes #9032 

If the agent is re-claimed at run-time then the new claim_id is not applied to the base-topic for the ACLK. This causes the broker to kick the agent for writing to the wrong topic.

##### Component Name
ACLK

##### Test Plan

* Start a claimed agent
* Remove the agent from its room
* Re-claim the agent (without stopping it), using -id=$(uuidgen)
* Re-connect the agent

##### Additional Information

underhood commented:
Hmm testing it got me into infinite reconnection loop of some kind :/

amoss commented:
It looks like the problem is caused by changing the claiming data while the link is operational. We should probably tear-down the connection if the claiming state is reloaded.

underhood commented:
Just for the record *(as i have memory of gold fish)*, I did skip the `Remove the agent from its room` step in test plan when this happened.

stelfrag commented on claim/claim.c:
```diff
@@ -109,12 +109,26 @@ void claim_agent(char *claiming_arguments)
 {
     if (claimed_id != NULL) {
         freez(claimed_id);
         claimed_id = NULL;
```
Very very small window that might cause a crash due to `char *agent_id = is_agent_claimed();` in `aclk_get_challenge` (L1242) if called in between the freez and the set to null.

amoss replied:
This is true and I've been wondering about it recently. The risk is minor but it is possible that accessing the cli-tool could trigger it.

hmoragrega commented on aclk/agent_cloud_link.c:
```diff
@@ -1319,6 +1333,8 @@ static void aclk_try_to_connect(char *hostname, char *port, int port_num)
     if (unlikely(rc)) {
         error("Failed to initialize the agent cloud link library");
     }
+    create_publish_base_topic();
```
I think this still has the issue that the agent will fail the first attempt to reconnect after a claiming id change,
the base topic (re) creation has to happen before the connection attempt at `mqtt_attempt_connection`, otherwise it will attempt to connect to the old outbound topic and verne will deny it, the second try will succeed

stelfrag replied:
I don't _see_ an issue with the base topic being created at that point (it can be moved a bit sure). 
The connection is down and it is trying to reconnect. The challenge / response always gets the fresh claim id (as long as the reload claiming has been done)

The base topic can be created after the connection attempt  (mqtt and lws upgrade) and only when the connection is marked UP it will attempt to subscribe to the `cmd` topic (with the newly created base topic)

hmoragrega replied:
the error happens because on connect we try to set the lwt in `_link_set_lwt` to the `outbound/meta` topic, and the first time we're still using the old `global_base_topic` with the previous id

amoss replied:
That's interesting. It worked well when I tested it because we don't use the topic for messages until later as @stelfrag mentioned. But I did not test the LWT and it will probably work the way that you have said. I will take another look.

amoss commented:
Rebased - new fix is a mutex around the `claim_id` to address @stelfrag 's concern, and I've moved the topic creation earlier to fix the problem with the LWT that @hmoragrega diagnosed.

amoss commented:
Fails in CI/CD - I need to check what is different to the build on my local system...

amoss commented:
CI/CD picked up a real problem - we had some existing code that would have failed the build when the cloud deps were not available. This has been in master for a while but not triggered the failure before. Fixed now.

underhood commented on web/api/web_api_v1.c:
```diff
@@ -874,10 +874,13 @@ inline int web_client_api_request_v1_info_fill_buffer(RRDHOST *host, BUFFER *wb)
-    if (is_agent_claimed() == NULL)
+    char *agent_id = is_agent_claimed();
+    if (agent_id == NULL)
         buffer_strcat(wb, "\t\"agent-claimed\": false,\n");
```
this is not the fault of this PR but I think we forgot to update the swagger docs with `agent-*` vars

amoss replied:
In theory...

underhood commented on aclk/agent_cloud_link.c:
```diff
@@ -429,29 +433,29 @@ struct aclk_query *aclk_queue_pop()
 char *create_publish_base_topic()
 {
-    if (unlikely(!is_agent_claimed()))
+char *agent_id = is_agent_claimed();
```
missing tab?

underhood commented on aclk/agent_cloud_link.c:
```diff
@@ -1577,13 +1607,13 @@ void aclk_disconnect()
 }
 
-void aclk_shutdown()
+/*void aclk_shutdown()
```
remove dead code?

underhood commented on claim/claim.c:
```diff
@@ -26,12 +26,19 @@ static char *claiming_errors[] = {
+char *is_agent_claimed()
 {
-    return claimed_id;
+char *result;
```
missing tab?

underhood commented on aclk/agent_cloud_link.c:
```diff
@@ -992,6 +996,39 @@ void *aclk_query_main_thread(void *ptr)
+
+    aclk_shutting_down = 1;
+    _link_shutdown();
+    aclk_lws_wss_mqtt_layer_disconect_notif();
```
shouldn't mosq_disconnect call this as well from `_link_shutdown`?

amoss replied:
Possibly, but this is a forced shutdown and I think the motivation was to make sure it had occurred.

underhood commented on aclk/agent_cloud_link.c:
```diff
@@ -992,6 +996,39 @@ void *aclk_query_main_thread(void *ptr)
+
+    event_loop_timeout = now_realtime_sec() + 5;
+    write_q = 1;
+    while (write_q && event_loop_timeout > now_realtime_sec()) {
```
this while is repeated few lines below - should we put it into func?

amoss replied:
No it is fine like this

underhood commented on claim/claim.c:
```diff
@@ -109,12 +116,35 @@ void claim_agent(char *claiming_arguments)
 
+#ifdef ENABLE_ACLK
+extern int aclk_connected, aclk_kill_link;
+extern void aclk_graceful_disconnect();
```
Is `aclk_graceful_disconnect` actually used here?

amoss replied:
No, that was from the previous version
